sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java

```diff
@@ -0,0 +1,317 @@
+                                upperBFinal,
+                                anAnalyticGroup.isRows,
+                                argList,
+                                combineFn,
+                                field);
+                        analyticFields.add(fieldAggregation);
+                      });
+            });
+
+    return new Transform(outputSchema, analyticFields);
+  }
+
+  private static class FieldAggregation implements Serializable {
+
+    private List<Integer> partitionKeys;
+    private List<Integer> orderKeys;
+    private List<Boolean> orderOrientations;
+    private List<Boolean> orderNulls;
+    private int lowerLimit = Integer.MAX_VALUE;
+    private int upperLimit = Integer.MAX_VALUE;
+    private boolean rows = true;
+    private List<Integer> inputFields;
+    private Combine.CombineFn combineFn;
+    private Schema.Field outputField;
+
+    public FieldAggregation(
+        List<Integer> partitionKeys,
+        List<Integer> orderKeys,
+        List<Boolean> orderOrientations,
+        List<Boolean> orderNulls,
+        int lowerLimit,
+        int upperLimit,
+        boolean rows,
+        List<Integer> inputFields,
+        Combine.CombineFn combineFn,
+        Schema.Field outputField) {
+      this.partitionKeys = partitionKeys;
+      this.orderKeys = orderKeys;
+      this.orderOrientations = orderOrientations;
+      this.orderNulls = orderNulls;
+      this.lowerLimit = lowerLimit;
+      this.upperLimit = upperLimit;
+      this.rows = rows;
+      this.inputFields = inputFields;
+      this.combineFn = combineFn;
+      this.outputField = outputField;
+    }
+  }
+
+  @Override
+  public NodeStats estimateNodeStats(RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    return inputStat;
+  }
+
+  @Override
+  public BeamCostModel beamComputeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    float multiplier = 1f + 0.125f;
+    return BeamCostModel.FACTORY.makeCost(
+        inputStat.getRowCount() * multiplier, inputStat.getRate() * multiplier);
+  }
+
+  private static class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+
+    private Schema outputSchema;
+    private List<FieldAggregation> aggFields;
+
+    public Transform(Schema schema, List<FieldAggregation> fieldAgg) {
+      this.outputSchema = schema;
+      this.aggFields = fieldAgg;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      PCollection<Row> inputData = input.get(0);
+      Schema inputSchema = inputData.getSchema();
+      for (FieldAggregation af : aggFields) {
+        if (af.partitionKeys.isEmpty()) {
+          // This sections simulate a KV Row
+          // Similar to the output of Group.byFieldIds
+          // When no partitions are specified
+          Schema inputSch = inputData.getSchema();
```

Comment:
The mock key generation here could be simplified by, something like

```
              windowedStream
                  .apply(WithKeys.of("dummy"))
                  .apply(GroupByKey.create())
```

note that maybe do not for many schema manipulation.